qa/rpc-tests/confidential_transactions.py

```diff
@@ -325,16 +329,122 @@ def run_test(self):
-        self.nodes[2].sendtoaddress(blinded_multisig_addr, 1)
+        # Issue new asset to check that partial blinding works with different assets in one transaction
+        issued3 = self.nodes[2].issueasset(1, 0)
+        self.nodes[2].generate(1)
         self.sync_all()
-        assert_equal(len(self.nodes[0].listunspent(0, 0, [unconfidential_addr])), 1)
-        assert_equal(len(self.nodes[1].listunspent(0, 0, [unconfidential_addr])), 1)
+        node2_balance = self.nodes[2].getbalance()
+        assert(issued3['asset'] in node2_balance)
+        assert_equal(node2_balance[issued3['asset']], Decimal(1))
+        # Send asset to blinded multisig address and check that they were received
+        self.nodes[2].sendtoaddress(blinded_multisig_addr, 1, "", "", False, issued3['asset'])
+        self.sync_all()
+        unspent_asset = self.nodes[0].listunspent(0, 0, [unconfidential_addr], True, issued3['asset'])
+        assert_equal(len(unspent_asset), 1)
+        assert(issued3['asset'] not in self.nodes[2].getbalance())
 
-        self.nodes[0].generate(1)
+        blinded_addr = self.nodes[0].getnewaddress()
+        blinded_addr2 = self.nodes[1].getnewaddress()
+        dst_addr = self.nodes[0].getnewaddress()
+        dst_addr2 = self.nodes[1].getnewaddress()
+        dst_addr3 = self.nodes[2].getnewaddress()
+        addr = self.nodes[0].validateaddress(blinded_addr)["unconfidential"]
+        addr2 = self.nodes[1].validateaddress(blinded_addr2)["unconfidential"]
+        self.nodes[0].sendtoaddress(blinded_addr, 0.1)
+        self.nodes[1].sendtoaddress(blinded_addr2, 0.11)
+
+        unspent = self.nodes[0].listunspent(0, 0, [addr])
+        unspent2 = self.nodes[1].listunspent(0, 0, [addr2])
+        assert_equal(len(unspent), 1)
+        assert_equal(len(unspent2), 1)
+
+        # Create one part of the transaction to partially blind
+        rawtx = self.nodes[0].createrawtransaction(
+            [{"txid": unspent2[0]["txid"], "vout": unspent2[0]["vout"]}], {dst_addr2: Decimal("0.01")})
+
+        # Create another part of the transaction to partially blind
+        rawtx2 = self.nodes[0].createrawtransaction(
+            [{"txid": unspent[0]["txid"], "vout": unspent[0]["vout"]},
+             {"txid": unspent_asset[0]["txid"], "vout": unspent_asset[0]["vout"]}],
+            {dst_addr: Decimal("0.1"), dst_addr3: Decimal("1.0")},
+            0,
+            {dst_addr: unspent[0]['asset'], dst_addr3: unspent_asset[0]['asset']})
+
+        sum_i = unspent2[0]["amount"] + unspent[0]["amount"]
+        sum_o = 0.01 + 0.10 + 0.1
+        assert_equal(int(round(sum_i*COIN)), int(round(sum_o*COIN)))
+
+        # Blind the first part of the transaction - we need to supply the
+        # assetcommmitments for all of the inputs, for the surjectionproof
+        # to be valid after we combine the transactions
+        blindtx = self.nodes[1].blindrawtransaction(
```

Comment:
wow I'm surprised it does the right thing here, even when the transaction itself only has 1 input! I guess 2 years ago me was thinking ahead...

Reply:
You note out of band that this is *indeed* broken behavior.

Could you please add this fix in `BlindTransaction` if it seems right:

```
     // Needed to construct the proof itself. Generators must match final transaction to be valid
     std::vector<secp256k1_generator> targetAssetGenerators;
     surjectionTargets.resize(tx.vin.size()*3);
+    if (auxiliary_generators) {
+        // It's fine to over-count here. We truncate before generating proofs.
+        surjectionTargets.resize(tx.vin.size()*3 + auxiliary_generators->size())
+    }
```

Reply:
Submitted #514 - made a little refactoring to avoid extra if(auxiliary_generators), and extra resize. Please check.